File: windows/electron/services/project-store.ts
import * as fs from 'fs'
import * as path from 'path'
import * as crypto from 'crypto'
import { v4 as uuidv4 } from 'uuid'
import { spawn } from 'child_process'

export type ProjectStatus =
  | 'active'
  | 'downloading'
  | 'uploading'
  | 'incloud'
  | 'archive'
  | 'ready'

export interface ProjectInfo {
  id: string
  name: string
  year?: string
  month?: string
  createdAt: string
  lastOpenedAt: string
  footageUrls: string[]
  status: ProjectStatus
  downloadProgress: Record<string, number>
  folderPath?: string
  sizeBytes?: number
}

const PROJECT_MANIFEST = '.edithub.json'
const PROJECT_METADATA = '.edithub-metadata.json'
export const DEFAULT_PROJECT_FOLDERS = [
  'FOOTAGE',
  'SFX',
  'MUSIC',
  'READY VIDEOS',
  'MISC',
  path.join('VOICE', 'ENCHANCE'),
  path.join('VOICE', 'NOT ENCHANCE'),
  'DOCS',
  'GRAPHICS',
  'B-ROLL',
  'SUBS',
]
const VIDEO_EXTENSIONS = new Set(['.mp4', '.mov', '.m4v', '.avi', '.mkv', '.webm', '.braw'])

export function readProjectInfo(projectFolder: string): ProjectInfo | null {
  try {
    const manifestPath = path.join(projectFolder, PROJECT_MANIFEST)
    if (!fs.existsSync(manifestPath)) return null
    const raw = fs.readFileSync(manifestPath, 'utf-8')
    const data = JSON.parse(raw) as ProjectInfo
    data.folderPath = projectFolder
    if (!data.year || !data.month) {
      const parts = projectFolder.split(path.sep)
      data.month = data.month || parts[parts.length - 2]
      data.year = data.year || parts[parts.length - 3]
    }
    return data
  } catch {
    return null
  }
}

function readJson(filePath: string): any | null {
  try {
    if (!fs.existsSync(filePath)) return null
    return JSON.parse(fs.readFileSync(filePath, 'utf-8'))
  } catch {
    return null
  }
}

function stableIdFromPath(projectFolder: string): string {
  return crypto.createHash('sha1').update(projectFolder.toLowerCase()).digest('hex').slice(0, 32)
}

function findMacManifest(projectFolder: string): any | null {
  const projectName = path.basename(projectFolder)
  return readJson(path.join(projectFolder, `${projectName}.edithub`))
}

function readMacProjectInfo(projectFolder: string, year?: string, month?: string): ProjectInfo | null {
  try {
    if (!fs.existsSync(projectFolder) || !fs.statSync(projectFolder).isDirectory()) return null

    const name = path.basename(projectFolder)
    const metadata = readJson(path.join(projectFolder, PROJECT_METADATA))
    const archiveManifest = findMacManifest(projectFolder)
    const stat = fs.statSync(projectFolder)

    const createdAt =
      stat.birthtime?.toISOString?.() ||
      stat.ctime.toISOString()

    return {
      id: metadata?.projectId || archiveManifest?.projectId || stableIdFromPath(projectFolder),
      name,
      year,
      month,
      createdAt,
      lastOpenedAt: stat.mtime.toISOString(),
      footageUrls: metadata?.footageLinks || archiveManifest?.footageLinks || [],
      status: archiveManifest ? 'archive' : 'ready',
      downloadProgress: {},
      folderPath: projectFolder,
    }
  } catch {
    return null
  }
}

export function writeProjectInfo(
  projectFolder: string,
  info: ProjectInfo
): void {
  const manifestPath = path.join(projectFolder, PROJECT_MANIFEST)
  fs.writeFileSync(manifestPath, JSON.stringify(info, null, 2), 'utf-8')
}

export function createProjectInfo(name: string, urls: string[]): ProjectInfo {
  return {
    id: uuidv4(),
    name,
    year: new Date().getFullYear().toString(),
    month: new Date().toLocaleString('en-US', { month: 'long' }).toUpperCase(),
    createdAt: new Date().toISOString(),
    lastOpenedAt: new Date().toISOString(),
    footageUrls: urls,
    status: urls.length > 0 ? 'downloading' : 'ready',
    downloadProgress: {},
  }
}

export function getFolderSizeBytes(folderPath: string): number {
  let total = 0
  try {
    const entries = fs.readdirSync(folderPath, { withFileTypes: true })
    for (const entry of entries) {
      const fullPath = path.join(folderPath, entry.name)
      if (entry.isDirectory()) {
        total += getFolderSizeBytes(fullPath)
      } else {
        try {
          total += fs.statSync(fullPath).size
        } catch {}
      }
    }
  } catch {}
  return total
}

export function extractZipProject(zipPath: string, destFolder: string): Promise<string | null> {
  return new Promise((resolve) => {
    try {
      const baseName = path.basename(zipPath, '.zip')
      const tempDest = path.join(destFolder, `__extracting_${baseName}_${Date.now()}`)
      fs.mkdirSync(tempDest, { recursive: true })

      const escapePsArg = (s: string) => `'${s.replace(/'/g, "''")}'`
      const ps = spawn('powershell.exe', [
        '-NoProfile', '-NonInteractive', '-Command',
        `Expand-Archive -LiteralPath ${escapePsArg(zipPath)} -DestinationPath ${escapePsArg(tempDest)} -Force`,
      ])

      ps.on('close', (code) => {
        if (code !== 0) {
          try { fs.rmSync(tempDest, { recursive: true, force: true }) } catch {}
          resolve(null)
          return
        }
        try { fs.unlinkSync(zipPath) } catch {}

        let finalPath: string
        try {
          const entries = fs.readdirSync(tempDest, { withFileTypes: true })
          const subDirs = entries.filter((e) => e.isDirectory())
          if (subDirs.length === 1) {
            const inner = path.join(tempDest, subDirs[0].name)
            finalPath = path.join(destFolder, subDirs[0].name)
            fs.renameSync(inner, finalPath)
            try { fs.rmdirSync(tempDest) } catch {}
          } else {
            finalPath = path.join(destFolder, baseName)
            fs.renameSync(tempDest, finalPath)
          }
        } catch {
          finalPath = tempDest
        }
        resolve(finalPath)
      })

      ps.on('error', () => {
        try { fs.rmSync(tempDest, { recursive: true, force: true }) } catch {}
        resolve(null)
      })
    } catch {
      resolve(null)
    }
  })
}

export function listProjectsInFolder(projectsFolder: string): ProjectInfo[] {
  const results: ProjectInfo[] = []
  const seen = new Set<string>()

  const pushProject = (info: ProjectInfo | null) => {
    if (!info) return
    if (seen.has(info.id)) return
    seen.add(info.id)
    results.push(info)
  }

  try {
    const scanProjectDir = (projectPath: string, year?: string, month?: string) => {
      const winInfo = readProjectInfo(projectPath)
      if (winInfo) {
        winInfo.folderPath = projectPath
        pushProject(winInfo)
        return
      }
      pushProject(readMacProjectInfo(projectPath, year, month))
    }

    const scanFolder = (folder: string, depth = 0, year?: string, month?: string) => {
      if (depth > 3) return
      const entries = fs.readdirSync(folder, { withFileTypes: true })
      for (const entry of entries) {
        if (
          entry.name.startsWith('.') ||
          entry.name === 'node_modules' ||
          entry.name.toLowerCase().startsWith('__extracting_')
        ) continue
        const entryPath = path.join(folder, entry.name)

        if (!entry.isDirectory()) continue

        const nextYear = /^\d{4}$/.test(entry.name) ? entry.name : year
        const nextMonth = year && depth <= 2 ? entry.name : month

        const hasWinManifest = fs.existsSync(path.join(entryPath, PROJECT_MANIFEST))
        const hasMacManifest = fs.existsSync(path.join(entryPath, `${entry.name}.edithub`))
        const hasMetadata = fs.existsSync(path.join(entryPath, PROJECT_METADATA))
        const hasKnownProjectFolders = ['SFX', 'Music', 'Footage', 'Graphics', 'Docs', 'FOOTAGE', 'READY VIDEOS', 'READY VIDEO', 'VOICE', 'B-ROLL', 'SUBS', 'MISC']
          .some((folderName) => fs.existsSync(path.join(entryPath, folderName)))

        if (hasWinManifest || hasMacManifest || hasMetadata || (year && month && hasKnownProjectFolders)) {
          scanProjectDir(entryPath, year, month)
          continue
        }

        scanFolder(entryPath, depth + 1, nextYear, nextMonth)
      }
    }

    if (fs.existsSync(projectsFolder)) {
      scanFolder(projectsFolder)
    }

    // Also support selecting the project folder itself.
    if (results.length === 0) {
      const own = readProjectInfo(projectsFolder) || readMacProjectInfo(projectsFolder)
      if (own) pushProject(own)
    }

  } catch (err) {
    console.warn('Error listing projects:', err)
  }
  return results.sort((a, b) => projectSortTime(b) - projectSortTime(a))
}

export function listArchivedProjectsInFolder(archiveFolder: string): ProjectInfo[] {
  const results: ProjectInfo[] = []
  try {
    if (!fs.existsSync(archiveFolder)) return []
    const walk = (dir: string, depth = 0) => {
      if (depth > 4) return
      const entries = fs.readdirSync(dir, { withFileTypes: true })
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name)
        if (entry.isDirectory()) {
          walk(fullPath, depth + 1)
          continue
        }
        if (!entry.name.endsWith('.zip')) continue
        const name = path.basename(entry.name, '.zip')
        const stat = fs.statSync(fullPath)
        const parts = path.relative(archiveFolder, fullPath).split(path.sep)
        const year = parts.length >= 3 ? parts[0] : ''
        const month = parts.length >= 3 ? parts[1] : ''
        results.push({
          id: stableIdFromPath(fullPath),
          name,
          year,
          month,
          createdAt: stat.birthtime.toISOString(),
          lastOpenedAt: stat.mtime.toISOString(),
          footageUrls: [],
          status: 'archive',
          downloadProgress: {},
          folderPath: fullPath,
          sizeBytes: stat.size,
        })
      }
    }
    walk(archiveFolder)
  } catch (err) {
    console.warn('Error listing archived projects:', err)
  }
  return results
}

export function createProjectFolderStructure(
  parentFolder: string,
  projectName: string
): string {
  const now = new Date()
  const year = now.getFullYear().toString()
  const month = now.toLocaleString('en-US', { month: 'long' }).toUpperCase()
  const projectFolder = path.join(parentFolder, year, month, projectName)
  const subfolders = DEFAULT_PROJECT_FOLDERS

  if (!fs.existsSync(projectFolder)) {
    fs.mkdirSync(projectFolder, { recursive: true })
  }

  for (const sub of subfolders) {
    const subPath = path.join(projectFolder, sub)
    if (!fs.existsSync(subPath)) {
      fs.mkdirSync(subPath, { recursive: true })
    }
  }

  return projectFolder
}

export interface FolderEntry {
  name: string
  path: string
  type: 'file' | 'folder'
  sizeBytes?: number
  children?: FolderEntry[]
}

export function listProjectFolders(projectFolder: string): FolderEntry[] {
  const readDir = (dir: string, depth = 0): FolderEntry[] => {
    if (!fs.existsSync(dir) || depth > 3) return []
    return fs.readdirSync(dir, { withFileTypes: true })
      .filter((entry) => !entry.name.startsWith('.'))
      .sort((a, b) => Number(b.isDirectory()) - Number(a.isDirectory()) || a.name.localeCompare(b.name))
      .map((entry) => {
        const fullPath = path.join(dir, entry.name)
        if (entry.isDirectory()) {
          return {
            name: entry.name,
            path: fullPath,
            type: 'folder' as const,
            children: readDir(fullPath, depth + 1),
          }
        }
        let sizeBytes = 0
        try { sizeBytes = fs.statSync(fullPath).size } catch {}
        return { name: entry.name, path: fullPath, type: 'file' as const, sizeBytes }
      })
  }
  return readDir(projectFolder)
}

export function findProjectPreviewVideo(projectFolder: string): string | null {
  const preferred = ['FOOTAGE', 'Footage', 'READY VIDEOS', 'READY VIDEO', 'Ready Videos']
  for (const folder of preferred) {
    const root = path.join(projectFolder, folder)
    const found = findFirstVideo(root)
    if (found) return found
  }
  return findFirstVideo(projectFolder)
}

const MONTH_INDEX: Record<string, number> = {
  JANUARY: 0,
  FEBRUARY: 1,
  MARCH: 2,
  APRIL: 3,
  MAY: 4,
  JUNE: 5,
  JULY: 6,
  AUGUST: 7,
  SEPTEMBER: 8,
  OCTOBER: 9,
  NOVEMBER: 10,
  DECEMBER: 11,
}

function projectSortTime(project: ProjectInfo): number {
  const year = Number(project.year)
  const monthKey = project.month?.toUpperCase()
  if (year && monthKey && monthKey in MONTH_INDEX) {
    const base = Date.UTC(year, MONTH_INDEX[monthKey], 1)
    const updated = new Date(project.lastOpenedAt || project.createdAt).getTime()
    return base + (Number.isFinite(updated) ? updated % (31 * 24 * 60 * 60 * 1000) : 0)
  }
  return new Date(project.lastOpenedAt || project.createdAt).getTime() || 0
}

export async function extractArchivesToProjects(
  sourceFolders: string[],
  onProgress: (p: { current: number; total: number; name: string }) => void
): Promise<{ extracted: number; total: number }> {
  interface ZipEntry { zipPath: string; year: string; month: string; name: string }

  const findZips = (dir: string, year?: string, month?: string): ZipEntry[] => {
    if (!fs.existsSync(dir)) return []
    let entries: fs.Dirent[]
    try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return [] }
    const results: ZipEntry[] = []
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name)
      if (entry.isDirectory()) {
        if (/^\d{4}$/.test(entry.name)) {
          results.push(...findZips(fullPath, entry.name, undefined))
        } else if (year && !month) {
          results.push(...findZips(fullPath, year, entry.name))
        }
      } else if (entry.name.endsWith('.zip') && year && month) {
        results.push({ zipPath: fullPath, year, month, name: path.basename(entry.name, '.zip') })
      }
    }
    return results
  }

  const all: ZipEntry[] = []
  const seen = new Set<string>()
  for (const folder of sourceFolders) {
    for (const z of findZips(folder)) {
      const key = `${z.year}/${z.month}/${z.name}`
      if (!seen.has(key)) { seen.add(key); all.push(z) }
    }
  }

  let extracted = 0
  for (let i = 0; i < all.length; i++) {
    const { zipPath, year, month, name } = all[i]
    onProgress({ current: i + 1, total: all.length, name })
    try {
      // Extract IN PLACE — into the same folder the zip lives in
      const destParent = path.dirname(zipPath)
      const expectedDest = path.join(destParent, name)
      if (fs.existsSync(expectedDest)) {
        try { fs.unlinkSync(zipPath) } catch {}
        continue
      }
      const result = await extractZipProject(zipPath, destParent)
      if (result && fs.existsSync(result)) {
        const manifestPath = path.join(result, '.edithub.json')
        if (!fs.existsSync(manifestPath)) {
          const stat = fs.statSync(result)
          const info: ProjectInfo = {
            id: stableIdFromPath(result),
            name: path.basename(result),
            year,
            month,
            createdAt: stat.birthtime?.toISOString?.() || stat.ctime.toISOString(),
            lastOpenedAt: stat.mtime.toISOString(),
            footageUrls: [],
            status: 'ready',
            downloadProgress: {},
          }
          writeProjectInfo(result, info)
        }
        extracted++
      }
    } catch (err) {
      console.warn(`Failed to extract ${zipPath}:`, err)
    }
  }

  return { extracted, total: all.length }
}

/**
 * Removes orphaned `__extracting_*` temp folders left behind when a zip
 * extraction crashed or the app was killed mid-extraction. A temp dir is
 * removed if either (a) a real sibling folder with the cleaned name already
 * exists (so the temp is a duplicate), or (b) it is older than `staleMs`
 * (an in-progress extraction always has a fresh mtime). Returns how many
 * dirs were removed and how many bytes were freed.
 */
export function sweepExtractingDirs(
  folders: string[],
  staleMs = 60 * 60 * 1000
): { removed: number; bytesFreed: number } {
  let removed = 0
  let bytesFreed = 0
  const now = Date.now()
  const seenRoots = new Set<string>()

  const cleanName = (name: string): string =>
    name.replace(/^__extracting_/i, '').replace(/_\d{10,}.*$/, '')

  const walk = (dir: string, depth = 0): void => {
    if (depth > 4 || !fs.existsSync(dir)) return
    let entries: fs.Dirent[]
    try { entries = fs.readdirSync(dir, { withFileTypes: true }) } catch { return }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue
      const fullPath = path.join(dir, entry.name)
      if (entry.name.toLowerCase().startsWith('__extracting_')) {
        const twin = path.join(dir, cleanName(entry.name))
        let isStale = false
        try { isStale = now - fs.statSync(fullPath).mtimeMs > staleMs } catch {}
        const hasTwin = fs.existsSync(twin) && twin !== fullPath
        if (hasTwin || isStale) {
          try {
            const size = getFolderSizeBytes(fullPath)
            fs.rmSync(fullPath, { recursive: true, force: true })
            if (!fs.existsSync(fullPath)) { removed++; bytesFreed += size }
          } catch (err) {
            console.warn(`Failed to remove stale extracting dir ${fullPath}:`, err)
          }
        }
        continue
      }
      walk(fullPath, depth + 1)
    }
  }

  for (const folder of folders) {
    const resolved = path.resolve(folder)
    if (seenRoots.has(resolved)) continue
    seenRoots.add(resolved)
    walk(resolved)
  }

  if (removed > 0) {
    console.log(`Swept ${removed} stale __extracting_ dir(s), freed ${(bytesFreed / 1e9).toFixed(2)} GB`)
  }
  return { removed, bytesFreed }
}

function findFirstVideo(root: string): string | null {
  if (!fs.existsSync(root)) return null
  try {
    const entries = fs.readdirSync(root, { withFileTypes: true })
    for (const entry of entries) {
      const fullPath = path.join(root, entry.name)
      if (entry.isFile() && VIDEO_EXTENSIONS.has(path.extname(entry.name).toLowerCase())) return fullPath
    }
    for (const entry of entries) {
      if (!entry.isDirectory()) continue
      const found = findFirstVideo(path.join(root, entry.name))
      if (found) return found
    }
  } catch {}
  return null
}

File: flutter_app/lib/services/project_store.dart
import 'dart:convert';
import 'dart:io';

import 'package:crypto/crypto.dart';
import 'package:path/path.dart' as p;
import 'package:uuid/uuid.dart';

import '../models/models.dart';

const String kProjectManifest = '.edithub.json';
const String kProjectMetadata = '.edithub-metadata.json';

const _archiveMonths = {
  '01': 'JANUARY',
  '02': 'FEBRUARY',
  '03': 'MARCH',
  '04': 'APRIL',
  '05': 'MAY',
  '06': 'JUNE',
  '07': 'JULY',
  '08': 'AUGUST',
  '09': 'SEPTEMBER',
  '10': 'OCTOBER',
  '11': 'NOVEMBER',
  '12': 'DECEMBER',
};

String archiveMonthFolder(String? month) {
  final value = month?.toUpperCase() ?? '';
  return _archiveMonths[value] ?? value;
}

bool _isArchiveMonth(String value) {
  final upper = value.toUpperCase();
  return _archiveMonths.containsKey(upper) ||
      _archiveMonths.containsValue(upper);
}

/// Subfolders created for every new project (matches the Electron app).
const List<String> kDefaultProjectFolders = [
  'FOOTAGE',
  'SFX',
  'MUSIC',
  'READY VIDEOS',
  'MISC',
  'VOICE/ENCHANCE',
  'VOICE/NOT ENCHANCE',
  'DOCS',
  'GRAPHICS',
  'B-ROLL',
  'SUBS',
];

const Set<String> kVideoExtensions = {
  '.mp4',
  '.mov',
  '.m4v',
  '.avi',
  '.mkv',
  '.webm',
  '.braw',
};

class ProjectStore {
  static const _uuid = Uuid();

  /// Stable id derived from the folder path — identical to the Electron app's
  /// `sha1(path.toLowerCase())[:32]` so manifests/ids line up.
  static String stableIdFromPath(String folder) {
    return sha1
        .convert(utf8.encode(folder.toLowerCase()))
        .toString()
        .substring(0, 32);
  }

  static Map<String, dynamic>? _readJson(String path) {
    try {
      final f = File(path);
      if (!f.existsSync()) return null;
      return jsonDecode(f.readAsStringSync()) as Map<String, dynamic>;
    } catch (_) {
      return null;
    }
  }

  /// Reads a Windows `.edithub.json` manifest, deriving year/month from the
  /// `{year}/{month}/{project}` path when the manifest omits them.
  ProjectInfo? readProjectInfo(String projectFolder) {
    final json = _readJson(p.join(projectFolder, kProjectManifest));
    if (json == null) return null;
    final info = ProjectInfo.fromJson(json)..folderPath = projectFolder;
    if (info.year == null || info.month == null) {
      final parts = p.split(projectFolder);
      final month =
          info.month ?? (parts.length >= 2 ? parts[parts.length - 2] : null);
      final year =
          info.year ?? (parts.length >= 3 ? parts[parts.length - 3] : null);
      return info.copyWith(year: year, month: month);
    }
    return info;
  }

  /// Reads a project archived by the Mac app (`{name}.edithub` /
  /// `.edithub-metadata.json`) or a footage-only folder.
  ProjectInfo? _readMacProjectInfo(
    String projectFolder,
    String? year,
    String? month,
  ) {
    try {
      final dir = Directory(projectFolder);
      if (!dir.existsSync()) return null;
      final name = p.basename(projectFolder);
      final metadata = _readJson(p.join(projectFolder, kProjectMetadata));
      final macManifest = _readJson(p.join(projectFolder, '$name.edithub'));
      final stat = dir.statSync();
      final footage =
          (metadata?['footageLinks'] ?? macManifest?['footageLinks'])
              as List? ??
          const [];
      return ProjectInfo(
        id:
            (metadata?['projectId'] ??
                    macManifest?['projectId'] ??
                    stableIdFromPath(projectFolder))
                as String,
        name: name,
        year: year,
        month: month,
        createdAt: stat.changed.toUtc().toIso8601String(),
        lastOpenedAt: stat.modified.toUtc().toIso8601String(),
        footageUrls: footage.map((e) => e.toString()).toList(),
        status: macManifest != null
            ? ProjectStatus.archive
            : ProjectStatus.ready,
        folderPath: projectFolder,
      );
    } catch (_) {
      return null;
    }
  }

  void writeProjectInfo(String projectFolder, ProjectInfo info) {
    File(
      p.join(projectFolder, kProjectManifest),
    ).writeAsStringSync(info.toManifestString());
  }

  ProjectInfo createProjectInfo(String name, List<String> urls) {
    final now = DateTime.now().toUtc().toIso8601String();
    return ProjectInfo(
      id: _uuid.v4(),
      name: name,
      year: DateTime.now().year.toString(),
      month: _monthName(DateTime.now().month),
      createdAt: now,
      lastOpenedAt: now,
      footageUrls: urls,
      status: urls.isEmpty ? ProjectStatus.ready : ProjectStatus.downloading,
    );
  }

  /// Scans `projectsFolder` (and nested year/month folders) for projects.
  List<ProjectInfo> listProjects(String projectsFolder) {
    final results = <ProjectInfo>[];
    final seen = <String>{};

    void push(ProjectInfo? info) {
      if (info == null || seen.contains(info.id)) return;
      seen.add(info.id);
      // Size is computed lazily per selected project in the detail view — never
      // walk every project's tree during a scan (that's a full-disk crawl).
      results.add(info);
    }

    void scanProjectDir(String path, String? year, String? month) {
      final win = readProjectInfo(path);
      if (win != null) {
        win.folderPath = path;
        push(win);
        return;
      }
      push(_readMacProjectInfo(path, year, month));
    }

    void scanFolder(String folder, int depth, String? year, String? month) {
      if (depth > 3) return;
      final dir = Directory(folder);
      if (!dir.existsSync()) return;
      for (final entry in dir.listSync(followLinks: false)) {
        if (entry is! Directory) continue;
        final name = p.basename(entry.path);
        if (name.startsWith('.') ||
            name == 'node_modules' ||
            name.toLowerCase().startsWith('__extracting_')) {
          continue;
        }
        final isYear = RegExp(r'^\d{4}$').hasMatch(name);
        final nextYear = isYear ? name : year;
        final nextMonth = year != null && _isArchiveMonth(name)
            ? archiveMonthFolder(name)
            : month;

        final hasWin = File(p.join(entry.path, kProjectManifest)).existsSync();
        final hasMac = File(p.join(entry.path, '$name.edithub')).existsSync();
        final hasMeta = File(p.join(entry.path, kProjectMetadata)).existsSync();
        if (hasWin || hasMac || hasMeta || (year != null && month != null)) {
          scanProjectDir(entry.path, year, month);
          continue;
        }
        scanFolder(entry.path, depth + 1, nextYear, nextMonth);
      }
    }

    scanFolder(projectsFolder, 0, null, null);
    results.sort((a, b) => _sortTime(b).compareTo(_sortTime(a)));
    return results;
  }

  /// Archived projects live as folders under `{icloud}/Videos/{year}/{month}`,
  /// so listing them is the same folder scan as active projects.
  List<ProjectInfo> listArchivedProjects(String archiveFolder) {
    return listProjects(
      archiveFolder,
    ).map((p) => p.copyWith(status: ProjectStatus.archive)).toList();
  }

  int getFolderSizeBytes(String folderPath) {
    var total = 0;
    try {
      final dir = Directory(folderPath);
      if (!dir.existsSync()) return 0;
      for (final entry in dir.listSync(recursive: true, followLinks: false)) {
        if (entry is File) {
          try {
            total += entry.lengthSync();
          } catch (_) {}
        }
      }
    } catch (_) {}
    return total;
  }

  /// Removes orphaned `__extracting_*` temp folders from interrupted zip
  /// extractions: a temp dir is dropped if a real sibling with the cleaned
  /// name exists, or if it is older than [staleAfter].
  ({int removed, int bytesFreed}) sweepExtractingDirs(
    List<String> folders, {
    Duration staleAfter = const Duration(hours: 1),
  }) {
    var removed = 0;
    var bytesFreed = 0;
    final now = DateTime.now();
    final seenRoots = <String>{};

    String cleanName(String name) => name
        .replaceFirst(RegExp(r'^__extracting_', caseSensitive: false), '')
        .replaceFirst(RegExp(r'_\d{10,}.*$'), '');

    void walk(String dir, int depth) {
      if (depth > 4) return;
      final d = Directory(dir);
      if (!d.existsSync()) return;
      List<FileSystemEntity> entries;
      try {
        entries = d.listSync(followLinks: false);
      } catch (_) {
        return;
      }
      for (final entry in entries) {
        if (entry is! Directory) continue;
        final name = p.basename(entry.path);
        if (name.toLowerCase().startsWith('__extracting_')) {
          final twin = p.join(dir, cleanName(name));
          var stale = false;
          try {
            stale = now.difference(entry.statSync().modified) > staleAfter;
          } catch (_) {}
          final hasTwin = twin != entry.path && Directory(twin).existsSync();
          if (hasTwin || stale) {
            try {
              final size = getFolderSizeBytes(entry.path);
              entry.deleteSync(recursive: true);
              removed++;
              bytesFreed += size;
            } catch (_) {}
          }
          continue;
        }
        walk(entry.path, depth + 1);
      }
    }

    for (final folder in folders) {
      final resolved = p.normalize(folder);
      if (seenRoots.contains(resolved)) continue;
      seenRoots.add(resolved);
      walk(resolved, 0);
    }
    return (removed: removed, bytesFreed: bytesFreed);
  }

  String createProjectFolderStructure(String parentFolder, String projectName) {
    final now = DateTime.now();
    final folder = p.join(
      parentFolder,
      now.year.toString(),
      _monthName(now.month),
      projectName,
    );
    if (Directory(folder).existsSync()) {
      throw FileSystemException(
        'A project with that name already exists.',
        folder,
      );
    }
    Directory(folder).createSync(recursive: true);
    for (final sub in kDefaultProjectFolders) {
      Directory(
        p.join(folder, sub.replaceAll('/', Platform.pathSeparator)),
      ).createSync(recursive: true);
    }
    return folder;
  }

  List<FolderEntry> listProjectFolders(String projectFolder) {
    List<FolderEntry> read(String dir, int depth) {
      final d = Directory(dir);
      if (!d.existsSync() || depth > 3) return [];
      final entries =
          d
              .listSync(followLinks: false)
              .where((e) => !p.basename(e.path).startsWith('.'))
              .toList()
            ..sort((a, b) {
              final ad = a is Directory ? 0 : 1;
              final bd = b is Directory ? 0 : 1;
              if (ad != bd) return ad - bd;
              return p.basename(a.path).compareTo(p.basename(b.path));
            });
      return entries.map((e) {
        final name = p.basename(e.path);
        if (e is Directory) {
          return FolderEntry(
            name: name,
            path: e.path,
            isFolder: true,
            children: read(e.path, depth + 1),
          );
        }
        int? size;
        try {
          size = (e as File).lengthSync();
        } catch (_) {}
        return FolderEntry(
          name: name,
          path: e.path,
          isFolder: false,
          sizeBytes: size,
        );
      }).toList();
    }

    return read(projectFolder, 0);
  }

  String? findProjectPreviewVideo(String projectFolder) {
    for (final folder in [
      'FOOTAGE',
      'Footage',
      'READY VIDEOS',
      'READY VIDEO',
      'Ready Videos',
    ]) {
      final found = _findFirstVideo(p.join(projectFolder, folder));
      if (found != null) return found;
    }
    return _findFirstVideo(projectFolder);
  }

  String? _findFirstVideo(String root) {
    final dir = Directory(root);
    if (!dir.existsSync()) return null;
    try {
      final entries = dir.listSync(followLinks: false);
      for (final e in entries) {
        if (e is File &&
            kVideoExtensions.contains(p.extension(e.path).toLowerCase())) {
          return e.path;
        }
      }
      for (final e in entries) {
        if (e is Directory) {
          final found = _findFirstVideo(e.path);
          if (found != null) return found;
        }
      }
    } catch (_) {}
    return null;
  }

  static const _months = [
    'JANUARY',
    'FEBRUARY',
    'MARCH',
    'APRIL',
    'MAY',
    'JUNE',
    'JULY',
    'AUGUST',
    'SEPTEMBER',
    'OCTOBER',
    'NOVEMBER',
    'DECEMBER',
  ];

  static String _monthName(int month) => _months[month - 1];

  int _sortTime(ProjectInfo project) {
    final year = int.tryParse(project.year ?? '');
    final monthIdx = project.month != null
        ? _months.indexOf(project.month!.toUpperCase())
        : -1;
    final updated = DateTime.tryParse(
      project.lastOpenedAt.isNotEmpty
          ? project.lastOpenedAt
          : project.createdAt,
    )?.millisecondsSinceEpoch;
    if (year != null && monthIdx >= 0) {
      final base = DateTime.utc(year, monthIdx + 1, 1).millisecondsSinceEpoch;
      const monthMs = 31 * 24 * 60 * 60 * 1000;
      return base + (updated != null ? updated % monthMs : 0);
    }
    return updated ?? 0;
  }
}
